### PR TITLE
Param requirements

### PR DESCRIPTION
**Pull request by Blacksmoke16**

### CODE REVIEW

**Review by drujensen — APPROVED** [review]
💯

**Review by elorest — APPROVED** [review]
That looks good.

File: README.md
Comment (robacarp) [inline README.md]: I’d love to see an added metric for the filters you’re adding here

File: README.md
Comment (robacarp) [inline README.md]: It’s interesting to see the speed difference here, which I’m assuming is due to change in crystal versions

**Comment by robacarp** [comment]
I have a couple thoughts, but I like the idea. I apparently can’t submit my comments on mobile... later I guess. Nice work though. 

Running ameba while also introducing changes makes it a bit hard to see he changes here. Would you mind breaking that out into a different pull?

**Comment by Blacksmoke16** [comment]
@robacarp Sure, can wait till you get a proper review in.

Most of the changes from ameba are specs.

Relevant files for this change would be:
* route_set.cr
* segment.cr
* routes_with_variables_spec.cr

Rest are ameba related only.

File: src/amber/router/segment.cr
Comment (Sija) [inline src/amber/router/segment.cr]:
```suggestion
    property pattern : Regex?
```

Seems more clear about its nature.

File: src/amber/router/route_set.cr
Comment (Sija) [inline src/amber/router/route_set.cr]: I'd add handling of requirements as `NamedTuple`.

**Review by robacarp — COMMENTED** [review]
Some of the code paths are a bit vague in my memory, so I may be overlooking something obvious. Is it possible to encapsulate everything around the idea of a requirement to be in the VariableSegment class? Then the #match? method would take care of storing the regex, the logic of checking to see if the parameter matches the regex, etc. What do you think?

File: README.md
Comment (Blacksmoke16) [reply README.md]: Either that or I have a beefier PC than what ran it previously.

**Comment by Blacksmoke16** [comment]
@robacarp Good call.  I added the pattern to only the variable segment class, and just handled the logic from within the `match?` method. 

I also added a section to the benchmark for a route with a requirement.  I should probably add one that fails to match now that i think of it... I'll do that when i get home.

@Sija I renamed the internal property to `pattern`, but left the param on the `add` method as `requirements` as it could include multiple patterns.

File: src/amber/router/route_set.cr
Comment (Blacksmoke16) [reply src/amber/router/route_set.cr]: How would you know the keys, since they are mapped to the named params in the route.

Comment (Sija) [reply src/amber/router/route_set.cr]: I'm not sure I understand. You can always transform it into `Hash` with `.to_h`.

**Comment by Sija** [comment]
Instead of `requirements` I'd suggest using `constraints`, which is more clear and commonly used term.

File: src/amber/router/route_set.cr
Comment (Blacksmoke16) [reply src/amber/router/route_set.cr]: Ohh i see what you mean.  Yea i can prob do that.  Was thinking you wanted the type restriction to be `NamedTuple` and not a hash.  My bad.

File: src/amber/router/segments/variable_segment.cr
Comment (robacarp) [inline src/amber/router/segments/variable_segment.cr]: This line is a bit hard to parse, can we make it a little more explicit?

```suggestion
      if pattern = @pattern
        segment =~ pattern
      else
        true
      end
```

**Comment by Blacksmoke16** [comment]
@robacarp @Sija This should be good to go now.

File: src/amber/router/route_set.cr
Comment (Sija) [reply src/amber/router/route_set.cr]: I belive this is still left.

**Comment by drujensen** [comment]
@robacarp @Sija need your approval

**Comment by Sija** [comment]
Just `NamedTuple` thing I mentioned in a comment before and it's GTG on my side.

**Comment by Blacksmoke16** [comment]
Ok how about now.

**Comment by Sija** [comment]
Why did you remove `ameba` dependency? Whatever the motivation you should open an issue for that since it's our of scope of this PR.

**Comment by Blacksmoke16** [comment]
It's implemented in #13

**Comment by Sija** [comment]
Oh, I see, fair enough!

**Review by Sija — APPROVED** [review]

**Review by robacarp — APPROVED** [review]
Thanks @Blacksmoke16, this looks great.

**Pull request merged**
